azure/linux-virtual-machine/variables.tf

```
@@ -208,3 +208,27 @@ variable "data_collection_rule_id" {
   type        = string
   default     = null
 }
+
```

This section would look better in a single object:

```suggestion
variable "autoshutdown" {
  description = "Describes the autoshutdown of this virtual machine"
  type = object({
    time = optional(string, "2200")
    timezone = optional(string, "UTC")
    email = optional(string)
  })
}
```
Doing it this way would eliminate the need for enable_autoshutdown entirely

---

I suppose what I'm trying to say is could you provide a use case where you would want to use:

```
autoshutdown = {
  enabled = false
}
```
rather than just excluding the block entirely?

If it's never going to be set to false via user interaction, then the user shouldn't need to see the variable at all

---

I can see it being treated similarly to `boot_diagnostics` where it being present, sure. Then if that's the case, we can have the description state "Set to an empty map to enable"?

I can check whether this will impact the actual module as we don't want it attempting to read values that don't yet exist, but if that's the case we can likely just replace the `var.x["key"]` with `lookup(var.x, "key")` and provide a default to satisfy whatever condition comes out of it.

---

Yes, we could have the description set as described, though this is a little redundant since the optional nature of everything in the object would imply that an empty map would enable it

As for the other variables, would this work?:

```
daily_recurrence_time = var.autoshutdown.time if var.autoshutdown
timezone              = = var.autoshutdown.timezone if var.autoshutdown
notification_settings {
  enabled = var.autoshutdown.email if var.autoshutdown
```

---

Just tested, setting `count = var.autoshutdown != null ? 1 : 0` seems to be adequate.

I'll update the branch soon, but just to track it:

- [ ] Update variable to be an object containing time, timezone and email with same defaults as original vars
- [ ] Update resource controlling schedule to define count based on `var.autoshutdown` being null or not
- [ ] Update description to "Describes the autoshutdown configuration with time being in 24h format and timezone being a supported timezone."

---

Updates pushed - ready for a new review now.